README.md
# Welcome to Babylon.js!

[Babylon](https://www.babylonjs.com/).

To get started, open a new shell and run:

```sh
$ npm init babylon@latest
```

Then follow the prompts you see in your terminal.

For more information about Babylon, [babylonjs.com](https://www.babylonjs.com/)!


CHANGELOG.md
# 2.0.2

fix: update to babylon 6.29.0
fix: vite 4.5

# 2.0.1

fix: update babylon to 6.2.0
fix: vite 4.3
fix: fix vite build to manual chunk babylon into own file

# 1.6.0

fix: change default layout to use top level async await
fix: update babylon

# 1.5.2

fix: update to vite 4
fix: update babylon

# 1.4.1

fix: vite 3 changed the default port to 5173

# 1.4.0

- update to Babylon 5.20.0
- update to Vite 3

# 1.3.0

- update to Babylon 5.10.0
- Update other deps

# 1.2.0

- update to Babylon 5.5.0
- Update other deps

# 1.1.0

- update to Babylon 5.1.0

# 1.0.7

- Update to Babylon 5.0.0-rc.1
- convert to ESM

# 1.0.0

- First version


cli.ts
#!/usr/bin/env node

import path from 'path'
import chalkAnimation from 'chalk-animation'
import inquirer from 'inquirer'
import meow from 'meow'

import { createApp } from './index.js'

const help = `
  Usage:
    $ npm init babylon [flags...] [<dir>]

  If <dir> is not provided up front you will be prompted for it.

  Flags:
    --help, -h          Show this help message
    --version, -v       Show the version of this script
`

run().then(
  () => {
    process.exit(0)
  },
  (error) => {
    console.error(error)
    process.exit(1)
  },
)

async function run() {
  let { input, flags, showHelp, showVersion } = meow(help, {
    importMeta: import.meta,
    flags: {
      help: { type: 'boolean', default: false, shortFlag: 'h' },
      version: { type: 'boolean', default: false, shortFlag: 'v' },
    },
  })

  if (flags.help) showHelp()
  if (flags.version) showVersion()

  let anim = chalkAnimation.rainbow(`\nBabylon\n`)
  await new Promise((res) => setTimeout(res, 1500))
  anim.stop()

  console.log("Welcome to Babylon! Let's get you set up with a new project.")
  console.log()

  // Figure out the app directory
  let projectDir = path.resolve(
    process.cwd(),
    input.length > 0
      ? input[0]
      : (
          await inquirer.prompt<{ dir: string }>([
            {
              type: 'input',
              name: 'dir',
              message: 'Where would you like to create your app?',
              default: './my-babylon-app',
            },
          ])
        ).dir,
  )

  let answers = await inquirer.prompt<{
    lang: 'ts' | 'js'
    install: boolean
  }>([
    // {
    //   name: 'lang',
    //   type: 'list',
    //   message: 'TypeScript or JavaScript?',
    //   choices: [
    //     { name: 'TypeScript', value: 'ts' },
    //     { name: 'JavaScript', value: 'js' },
    //   ],
    // },
    {
      name: 'install',
      type: 'confirm',
      message: 'Do you want me to run `npm install`?',
      default: true,
    },
  ])

  await createApp({
    projectDir,
    lang: answers.lang || 'ts',
    install: answers.install,
  })
}


index.ts
import path from 'path'
import { execSync } from 'child_process'
import fse from 'fs-extra'
import { URL } from 'url'

export type Lang = 'ts' | 'js'

interface CreateAppArgs {
  projectDir: string
  lang: Lang
  install: boolean
}

async function createApp({ projectDir, lang, install }: CreateAppArgs) {
  // Create the app directory
  let relativeProjectDir = path.relative(process.cwd(), projectDir)
  let projectDirIsCurrentDir = relativeProjectDir === ''
  if (!projectDirIsCurrentDir) {
    if (fse.existsSync(projectDir)) {
      console.log(
        `️🚨 Oops, "${relativeProjectDir}" already exists. Please try again with a different directory.`,
      )
      process.exit(1)
    } else {
      await fse.mkdir(projectDir)
    }
  }

  // copy the template
  const template = new URL(`../templates/${lang}`, import.meta.url).pathname
  await fse.copy(template, projectDir)

  if (install) {
    console.log('📦 Installing dependencies...')
    execSync('npm install', { stdio: 'inherit', cwd: projectDir })
  }

  if (projectDirIsCurrentDir) {
    console.log(`That's it! Check the README for development and deploy instructions!`)
  } else {
    console.log(
      `That's it! \`cd\` into "${relativeProjectDir}" and check the README for development and deploy instructions!`,
    )
  }
}

export { createApp }


templates/ts/src/main.ts
import './style.css'

import { ArcRotateCamera } from '@babylonjs/core/Cameras/arcRotateCamera'
import { Engine } from '@babylonjs/core/Engines/engine'
import { SceneLoader } from '@babylonjs/core/Loading/sceneLoader'
import { ImageProcessingConfiguration } from '@babylonjs/core/Materials/imageProcessingConfiguration'
import { Vector3 } from '@babylonjs/core/Maths/math.vector'
import { Scene } from '@babylonjs/core/scene'
import { createEnvironment } from './helpers'

import '@babylonjs/loaders/glTF/2.0' // needed for gltf 2

// Engine
const canvas = document.getElementById('app') as HTMLCanvasElement
const engine = new Engine(canvas, true, {}, true)
const handleResize = () => engine.resize()
window.addEventListener('resize', handleResize, { passive: true })

// Scene
const scene = new Scene(engine)
scene.imageProcessingConfiguration.toneMappingEnabled = true
scene.imageProcessingConfiguration.toneMappingType = ImageProcessingConfiguration.TONEMAPPING_ACES

// Environment
const skyBox = createEnvironment(scene)

// Camera
const startAlpha = (Math.PI / 4) * 3 // rotate so we see the cube
const startBeta = Math.PI / 3 // look from above
const camera = new ArcRotateCamera('arcRotateCamera', startAlpha, startBeta, 10, Vector3.Zero(), scene)
scene.switchActiveCamera(camera)

// Load GLTF Model
const container = await SceneLoader.LoadAssetContainerAsync('/models/', 'babylon-cube.glb', scene)
container.addAllToScene()

// Calculate world size
const worldExtent = scene.getWorldExtends()
const worldSize = worldExtent.max.subtract(worldExtent.min)
const worldCenter = worldExtent.max.add(worldExtent.min).scale(0.5)

// position camera at center
camera.setTarget(worldCenter)
camera.radius = worldSize.length() * 1.5

// Start Render Loop
engine.runRenderLoop(() => scene.render())


templates/ts/src/babylon.ts
import {
  Engine,
  Scene,
  ArcRotateCamera,
  Vector3,
  CubeTexture,
  Color4,
  ImageProcessingConfiguration,
} from '@babylonjs/core'

export let canvas: HTMLCanvasElement
export let engine: Engine
export let scene: Scene
export let camera: ArcRotateCamera
let handleResize: any

export const createEngine = (hostCanvas: HTMLCanvasElement) => {
  canvas = hostCanvas
  engine = new Engine(canvas, true, {}, true)

  handleResize = () => engine.resize()
  window.addEventListener('resize', handleResize)

  return engine
}

export const createScene = () => {
  scene = new Scene(engine)

  scene.clearColor = new Color4(0.8, 0.8, 0.8, 1)

  // optimize scene for opaque background
  scene.autoClear = false
  scene.autoClearDepthAndStencil = false

  // setup ACES tone mapping for more natural colors
  scene.imageProcessingConfiguration.toneMappingEnabled = true
  scene.imageProcessingConfiguration.toneMappingType = ImageProcessingConfiguration.TONEMAPPING_ACES

  // show the inspector when pressing shift + alt + I
  // install @babylonjs/inspector and import here
  // scene.onKeyboardObservable.add(({ event }) => {
  //   if (event.ctrlKey && event.shiftKey && event.code === 'KeyI') {
  //     if (scene.debugLayer.isVisible()) {
  //       scene.debugLayer.hide()
  //     } else {
  //       scene.debugLayer.show()
  //     }
  //   }
  // })

  return scene
}

export const createArcRotateCamera = () => {
  const startAlpha = -(Math.PI / 2)
  const startBeta = Math.PI / 2
  const startRadius = 10
  const startPosition = new Vector3(0, 0, 0)
  const camera = new ArcRotateCamera(
    'camera',
    startAlpha,
    startBeta,
    startRadius,
    startPosition,
    scene,
    true,
  )
  camera.attachControl(canvas, false)

  // Set some basic camera settings
  camera.minZ = 1 // clip at 1 meter

  camera.panningAxis = new Vector3(1, 0, 1) // pan along ground
  camera.panningSensibility = 1000 // how fast do you pan, set to 0 if you don't want to allow panning
  camera.panningOriginTarget = Vector3.Zero() // where does the panning distance limit originate from
  camera.panningDistanceLimit = 100 // how far can you pan from the origin

  camera.allowUpsideDown = false // don't allow zooming inverted
  camera.lowerRadiusLimit = 2 // how close can you zoom
  camera.upperRadiusLimit = 100 // how far out can you zoom
  camera.lowerBetaLimit = 0.5 // how high can you move the camera
  camera.upperBetaLimit = 1.4 // how low down can you move the camera

  camera.checkCollisions = true // make the camera collide with meshes
  camera.collisionRadius = new Vector3(1, 1, 1) // how close can the camera go to other meshes

  return camera
}

export const createPBRSkybox = () => {
  const environmentTexture = CubeTexture.CreateFromPrefilteredData(
    '/environments/environment-specular.env',
    scene,
  )

  const skyboxMesh = scene.createDefaultSkybox(environmentTexture, true, 1000, 0.5, true)

  return skyboxMesh
}


templates/ts/src/helpers.ts
import { BackgroundMaterial } from '@babylonjs/core/Materials/Background/backgroundMaterial'
import { CubeTexture } from '@babylonjs/core/Materials/Textures/cubeTexture'
import { Texture } from '@babylonjs/core/Materials/Textures/texture'
import { CreateBox } from '@babylonjs/core/Meshes/Builders/boxBuilder'
import { Scene } from '@babylonjs/core/scene'

import '@babylonjs/core/Materials/Textures/Loaders/envTextureLoader' // needed for environment texture
import '@babylonjs/core/Misc/dds' // needed for environment texture

export function createEnvironment(scene: Scene) {
  const environmentTexture = CubeTexture.CreateFromPrefilteredData('/environments/environment-specular.env', scene)
  scene.environmentTexture = environmentTexture

  const skyBox = CreateBox('skyBox', { size: 1000 }, scene)
  skyBox.isPickable = false
  skyBox.infiniteDistance = true
  skyBox.ignoreCameraMaxZ = true
  const skyBoxMaterial = new BackgroundMaterial('skyBoxMaterial', scene)
  skyBox.material = skyBoxMaterial
  skyBoxMaterial.reflectionTexture = environmentTexture.clone()
  skyBoxMaterial.reflectionTexture.coordinatesMode = Texture.SKYBOX_MODE
  skyBoxMaterial.backFaceCulling = false
  skyBoxMaterial.reflectionBlur = 0.5
  skyBoxMaterial.opacityFresnel = false
}


templates/ts/src/runtime.ts
import { ArcRotateCamera } from '@babylonjs/core/Cameras/arcRotateCamera'
import { Engine } from '@babylonjs/core/Engines/engine'
import { ImageProcessingConfiguration } from '@babylonjs/core/Materials/imageProcessingConfiguration'
import { CubeTexture } from '@babylonjs/core/Materials/Textures/cubeTexture'
import { Color4 } from '@babylonjs/core/Maths/math.color'
import { Vector3 } from '@babylonjs/core/Maths/math.vector'
import { Scene } from '@babylonjs/core/scene'
import '@babylonjs/core/Helpers/sceneHelpers' // needed to augment scene with createDefaultSkybox
import '@babylonjs/core/Collisions/collisionCoordinator' // needed so camera can collide with meshes

export let canvas: HTMLCanvasElement
export let engine: Engine
export let scene: Scene
export let camera: ArcRotateCamera

export const createEngine = (hostCanvas: HTMLCanvasElement) => {
  canvas = hostCanvas
  engine = new Engine(canvas, true, {}, true)

  const handleResize = () => engine.resize()
  window.addEventListener('resize', handleResize)
  window.addEventListener('unload', () => window.removeEventListener('resize', handleResize), {
    once: true,
  })

  return engine
}

export const createScene = () => {
  scene = new Scene(engine)

  // change to a neutral color instead of babylon purple default
  scene.clearColor = new Color4(0.8, 0.8, 0.8, 1)

  // optimize scene for opaque background
  scene.autoClear = false
  scene.autoClearDepthAndStencil = false

  // setup ACES tone mapping for more natural colors
  scene.imageProcessingConfiguration.toneMappingEnabled = true
  scene.imageProcessingConfiguration.toneMappingType = ImageProcessingConfiguration.TONEMAPPING_ACES

  // show the inspector when pressing shift + alt + I
  // install @babylonjs/inspector and import here
  // scene.onKeyboardObservable.add(({ event }) => {
  //   if (event.ctrlKey && event.shiftKey && event.code === 'KeyI') {
  //     if (scene.debugLayer.isVisible()) {
  //       scene.debugLayer.hide()
  //     } else {
  //       scene.debugLayer.show()
  //     }
  //   }
  // })

  return scene
}

export const createArcRotateCamera = () => {
  const startAlpha = (Math.PI / 4) * 3 // horizontal rotation
  const startBeta = Math.PI / 4 // vertical angle
  const startRadius = 5 // how far from center
  const startPosition = new Vector3(0, 0, 0)
  const camera = new ArcRotateCamera(
    'camera',
    startAlpha,
    startBeta,
    startRadius,
    startPosition,
    scene,
    true,
  )
  camera.attachControl(canvas, false)

  // Set some basic camera settings
  camera.minZ = 1 // clip at 1 meter

  camera.panningAxis = new Vector3(1, 0, 1) // pan along ground
  camera.panningSensibility = 1000 // how fast do you pan, set to 0 if you don't want to allow panning
  camera.panningOriginTarget = Vector3.Zero() // where does the panning distance limit originate from
  camera.panningDistanceLimit = 100 // how far can you pan from the origin

  camera.allowUpsideDown = false // don't allow zooming inverted
  camera.lowerRadiusLimit = 2 // how close can you zoom
  camera.upperRadiusLimit = 100 // how far out can you zoom
  camera.lowerBetaLimit = 0.5 // how high can you move the camera
  camera.upperBetaLimit = 1.4 // how low down can you move the camera

  camera.checkCollisions = true // make the camera collide with meshes
  camera.collisionRadius = new Vector3(1, 1, 1) // how close can the camera go to other meshes

  camera.useAutoRotationBehavior = true // rotate around the target

  return camera
}

export const createPBRSkybox = () => {
  const environmentTexture = CubeTexture.CreateFromPrefilteredData(
    '/environments/environment-specular.env',
    scene,
  )
  return scene.createDefaultSkybox(environmentTexture, true, 150, 0.5, true)
}
